apps/openmw/mwrender/objects.cpp

```diff
@@ -66,22 +184,125 @@ void Objects::insertBegin(const MWWorld::Ptr& ptr)
     insert->setScale(scaleVec);
 
     ptr.getRefData().setBaseNode(insert);
+    return cellnode;
 }
 
+struct OctreeAddRemove
+{
+    OctreeAddRemove(const OcclusionQuerySettings & settings): mSettings(settings) {}
+    const OcclusionQuerySettings & mSettings;
+
+    void recursivCellAddStaticObject(osg::BoundingSphere&bs, osg::OcclusionQueryNode &parent, osg::Group *child, osg::BoundingSphere& childbs)
+    {
+        osg::Vec3i index =osg::Vec3i(childbs.center()[0]<bs.center()[0]?0:1,
+                                     childbs.center()[1]<bs.center()[1]?0:1,
+                                     childbs.center()[2]<bs.center()[2]?0:1);
+        uint ind = index[0] + index[1]*2 + index[2]*4;
+        osg::Vec3 indexf = osg::Vec3(index[0]*2-1, index[1]*2-1, index[2]*2-1);
+        osg::BoundingSphere bsi;
+        bsi.radius() = bs.radius() * 0.5f;
+        bsi.center() = bs.center() + indexf*bsi.radius();
+        osg::ref_ptr<osg::OcclusionQueryNode> qnode;
+        osg::ref_ptr<osg::Group> target = parent.getChild(ind)->asGroup();
+
+        osgUtil::GeometryCollector geomcollector(0,osgUtil::Optimizer::ALL_OPTIMIZATIONS);
+        target->accept(geomcollector);
+
+        if( target->getBound().radius() > mSettings.minOQNSize
+            && bsi.radius() > mSettings.minOQNSize
+            && geomcollector.getGeometryList().size() > mSettings.maxDrawablePerOQN
+          )
+        {
+            qnode=dynamic_cast<osg::OcclusionQueryNode*>(target.get());
+            if(!qnode.valid())
+            {
+                OSG_INFO<<"new OcclusionQueryNode with radius "<<bs.radius()<<std::endl;
+                qnode= new HierarchicalBoundingOcclusionQueryNode;
+                qnode->setVisibilityThreshold(parent.getVisibilityThreshold());
+                qnode->setDebugDisplay(parent.getDebugDisplay());
+                qnode->setQueryFrameCount(parent.getQueryFrameCount());
+                for(uint i=0; i<8; ++i)
+                    qnode->addChild(new osg::Group());
+
+                for(uint i=0; i<target->getNumChildren(); ++i)
+                {
+                    osg::Group * child = target->getChild(i)->asGroup();
+                    osg::BoundingSphere bschild = child->getBound();
+                    recursivCellAddStaticObject(bsi, *qnode, child, bschild);
+                }
+                parent.setChild(ind, qnode);
+            }
+            recursivCellAddStaticObject(bsi, *qnode, child, childbs);
+
+        }
+        else
+        {
+            target->addChild(child);
+        }
+
+    }
+    bool recursivCellRemoveStaticObject(osg::OcclusionQueryNode & parent, osg::Node * childtoremove)
+    {
+        osg::Group * pchild; bool removed=false;
+        for(uint i=0; i< parent.getNumChildren(); ++i)
+        {
+            pchild=(parent.getChild(i)->asGroup());
+            if(removed = pchild->removeChild(childtoremove)) break;
```

Comment:
I suppose there should be comparison, not assignment.

Reply:
no ..i'll put parenthesis
...this breaks save feature will see that

Reply:
> Does not compile with OSG 3.6.0 - it complains about undeclared `_validQueryGeometry` variable and missing `osg::QueryGeometry::QueryResult` type.

this require lattest release candidate of 3.6...changes have been made two month ago